lib/models/addon-discovery.js

```diff
@@ -20,241 +20,237 @@ var findup     = require('findup');
+    It returns a tree of plain objects that each contain information about a
+    discovered addon. Each node has `name`, `path`, `pkg` and
+    `childAddons` properties. The latter is an array containing any addons
+    discovered from applying the discovery process to that addon.
+
+    @private
+    @method discoverProjectAddons
+   */
+  discoverProjectAddons: function(project) {
+    var projectAsAddon = this.discoverFromProjectItself(project);
+    var internalAddons = this.discoverFromInternalProjectAddons(project);
+    var cliAddons = this.discoverFromCli(project.cli);
+    var dependencyAddons;
+
+    if (project.hasDependencies()) {
+      dependencyAddons = this.discoverFromDependencies(project.root, project.nodeModulesPath, project.pkg, false);
+    }  else {
+      dependencyAddons = [];
     }
-  }
-  return [];
-};
 
-/**
-  Returns a tree based on the addons referenced in the provided `pkg` through
-  the package.json `dependencies` and optionally `devDependencies` collections,
-  as well as those discovered addons' child addons.
-
-  @private
-  @method discoverFromDependencies
- */
-AddonDiscovery.prototype.discoverFromDependencies = function(root, nodeModulesPath, pkg, excludeDevDeps) {
-  var discovery = this;
-  var addons = Object.keys(this.dependencies(pkg, excludeDevDeps)).map(function(name) {
-    if (name !== 'ember-cli') {
-      var addonPath = this.resolvePackage(root, name);
-
-      if (addonPath) {
-        return discovery.discoverAtPath(addonPath);
-      }
-
-      // this supports packages that do not have a valid entry point
-      // script (aka `main` entry in `package.json` or `index.js`)
-      addonPath = path.join(nodeModulesPath, name);
-      var addon = discovery.discoverAtPath(addonPath);
-      if (addon) {
-        var chalk = require('chalk');
-
-        discovery.ui.writeLine(chalk.yellow('The package `' + name + '` is not a properly formatted package, we have used a fallback lookup to resolve it at `' + addonPath + '`. This is generally caused by an addon not having a `main` entry point (or `index.js`).'), 'WARNING');
-
-        return addon;
+    var inRepoAddons = this.discoverInRepoAddons(project.root, project.pkg);
+    var addons = projectAsAddon.concat(cliAddons, internalAddons, dependencyAddons, inRepoAddons);
+
+    return addons;
+  },
+
+  /**
+    This is one of the primary APIs for this class and is called by addons.
+    It returns a tree of plain objects that each contain information about a
+    discovered addon. Each node has `name`, `path`, `pkg` and
+    `childAddons` properties. The latter is an array containing any addons
+    discovered from applying the discovery process to that addon.
+
+    @private
+    @method discoverProjectAddons
+  */
+  discoverChildAddons: function(addon) {
+    debug('discoverChildAddons: %s(%s)', addon.name, addon.root);
+    var dependencyAddons = this.discoverFromDependencies(addon.root, addon.nodeModulesPath, addon.pkg, true);
```

Comment:
> do we need a addon.hasDependencies check?

the original code does not, if we need to another PR should likely tackle



> also docs are copy pasted

docs are quite similar, but not the same. Also this is pre-existing, merely transposed into the newer form.